### PR TITLE
[lua] more lsp backends for lua

### PR DESCRIPTION
**Pull request by sunlin7**
Add more lua lsp backends support, and keep compatible with previous `lsp-emmy`.

### CODE REVIEW

File: layers/+lang/lua/funcs.el [hunk 1]
```diff
@@ -23,8 +29,11 @@
     "sl" 'lua-send-current-line
     "sr" 'lua-send-region
     "'" 'lua-show-process-buffer)
-  (pcase lua-backend
-    (`lsp-emmy (spacemacs//lua-setup-lsp-emmy))))
+  (pcase (spacemacs//lua-backend)
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:32]: Multiple issues in the `pcase` form.

1. As mentioned above, the function call to `spacemacs//lua-backend` should be inlined here.
 the problem is, one of the allowed value, `lsp-lua`, is not mentioned in the docstring of `lua-backend`.
2. The branch of `lsp-lua` calls `spacemacs//lua-setup-lsp-lua`, but this doesn't make sense at all because it doesn't select any particular backend. So this entire branch, including the function `spacemacs//lua-setup-lsp-lua` should be removed.
3. Then it becomes clear that the entire `spacemacs//lua-backend` is redundant and does nothing.
4. In all branches, the name of the backend could simply be quoted by `'` instead of backquote.

So this `pcase` form should be changed to the following
```elisp
(pcase lua-backend
   ('lsp-emmy-lua (spacemacs//lua-setup-lsp-emmy-lua))
   ('lsp-lua-language-server (spacemacs//lua-setup-lsp-lua-language-server))
   ('lsp-lua-lsp (spacemacs//lua-setup-lsp-lua-lsp)))
```

File: layers/+lang/lua/funcs.el [hunk 2]
```diff
@@ -33,19 +42,34 @@
...
-(defun spacemacs//lua-setup-lsp-emmy ()
-  "Setup LSP Lua."
+(defun spacemacs//lua-setup-lsp-emmy-lua ()
+  "Setup emmy Lua."
   (when lua-lsp-emmy-java-path
     (setq lsp-clients-emmy-lua-java-path lua-lsp-emmy-java-path))
   (when lua-lsp-emmy-jar-path
     (setq lsp-clients-emmy-lua-jar-path (expand-file-name lua-lsp-emmy-jar-path)))
   (setq lsp-enable-file-watchers lua-lsp-emmy-enable-file-watchers)
+  (set (make-local-variable 'lsp-disabled-clients) '(lua-language-server lsp-lua-lsp))
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:52]: This should be replaced by `setq-local`:
```elisp
(setq-local lsp-disabled-clients '(lua-language-server lsp-lua-lsp))
```

File: layers/+lang/lua/funcs.el [hunk 3]
```diff
@@ -33,19 +42,34 @@
...
+  (lsp))
+
+(defun spacemacs//lua-setup-lsp-lua-language-server ()
+  "Setup lua-language-server."
+  (set (make-local-variable 'lsp-disabled-clients) '(emmy-lua lsp-lua-lsp))
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:57]: Same as above.

File: layers/+lang/lua/funcs.el [hunk 4]
```diff
@@ -33,19 +42,34 @@
...
   (lsp))
 
+(defun spacemacs//lua-setup-lsp-lua-lsp ()
+  "Setup lsp-lua-lsp."
+  (set (make-local-variable 'lsp-disabled-clients) '(emmy-lua lsp-lua-language-server))
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:62]: Same as above.

File: layers/+lang/lua/funcs.el [hunk 5]
```diff
@@ -33,19 +42,34 @@
...
+(defun spacemacs//lua-setup-lsp-lua ()
+  "Setup lsp for lua"
+  (lsp))
 

 ;; Lua mode
 (defun spacemacs//lua-setup-company-lua ()
   (spacemacs|add-company-backends
     :backends company-lua
-    :modes lua-mode)
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:50]: It's not clear to me why these changes are made.

**Review by lebensterben — CHANGES_REQUESTED** [review]
Please see my comments.

File: layers/+lang/lua/README.org [hunk 6]
```diff
@@ -36,24 +36,35 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
 
 #+BEGIN_SRC elisp
   (lua :variables
        lua-backend 'lsp-emmy
-       lua-lsp-emmy-jar-path "~/.emacs.d/EmmyLua-LS-all.jar" ; default path
-       lua-lsp-emmy-java-path "java"                         ; default path
-       lua-lsp-emmy-enable-file-watchers t)                  ; enabled default
+       lsp-clients-emmy-lua-jar-path "~/.emacs.d/.cache/lsp/EmmyLua-LS-all.jar" ; default path
+       lsp-clients-emmy-lua-java-path "java" ; default path
+       lsp-enable-file-watchers t)           ; enabled default
```
[the remark this answers is not in the document]
Comment (sunlin7) [reply layers/+lang/lua/README.org]: resolved in new submition.

File: layers/+lang/lua/README.org [hunk 7]
```diff
@@ -36,24 +36,35 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
 
-To use the =lsp-emmy= backend, please download the [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua-LS-all.jar]] and put it into =~/.emacs.d/=,
-then set the layer variable lua-backend to =lsp-emmy=.
+To use the =lsp-emmy-lua= backend, please download the [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua-LS-all.jar]] and put it into =~/.emacs.d/=,
+then set the layer variable lua-backend to =lsp-emmy-lua=.
 
-=Lsp-emmy= requires a working java installation in the path as well as the path to the LSP server jar.
-The java path and the path to the server executable can be configured as is shown below:
+=lsp-emmy-lua= requires a working java installation in the path as well as the path to the LSP server jar.
+The java path and the path to the server executable can be configured as is shown bellow:
```
[the remark this answers is not in the document]
Comment (sunlin7) [reply layers/+lang/lua/README.org]: resolved in new submition.

File: layers/+lang/lua/README.org [hunk 8]
```diff
@@ -36,24 +36,35 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
-       lua-lsp-emmy-jar-path "~/.emacs.d/EmmyLua-LS-all.jar" ; default path
-       lua-lsp-emmy-java-path "java"                         ; default path
-       lua-lsp-emmy-enable-file-watchers t)                  ; enabled default
+       lsp-clients-emmy-lua-jar-path "~/.emacs.d/.cache/lsp/EmmyLua-LS-all.jar" ; default path
+       lsp-clients-emmy-lua-java-path "java" ; default path
+       lsp-enable-file-watchers t)           ; enabled default
+#+END_SRC
+
+To use the =lsp-lua-language-server= backend, please following [[https://github.com/sumneko/lua-language-server][lua-language-server]] to compile the standalone binnary,
+then set the layer variable lua-backend to =lsp-lua-language-server= and the variables for binnary and main entry, an example as bellow:
```
[the remark this answers is not in the document]
Comment (sunlin7) [reply layers/+lang/lua/README.org]: resolved in new submition.

File: layers/+lang/lua/config.el [hunk 9]
```diff
@@ -13,15 +13,22 @@
 
 (spacemacs|define-jump-handlers lua-mode)
 
-(defvar lua-backend 'nil
-  "The backend to be used for lua must be `lsp-emmy' or nil if `lua-mode' should be used.")
+(defcustom lua-backend nil
```
[the remark this answers is not in the document]
Comment (sunlin7) [reply layers/+lang/lua/config.el]: resolved in new submition.

File: layers/+lang/lua/funcs.el (hunk 1 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: resolved in new submition.

File: layers/+lang/lua/funcs.el (hunk 3 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: resolved in new submition.

File: layers/+lang/lua/funcs.el (hunk 2 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: resolved in new submition.

File: layers/+lang/lua/funcs.el (hunk 4 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: resolved in new submition.

File: layers/+lang/lua/funcs.el (hunk 5 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: resolved in new submition.

File: layers/+lang/lua/README.org [hunk 10]
```diff
@@ -36,24 +36,35 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
        lua-backend 'lsp-emmy
-       lua-lsp-emmy-jar-path "~/.emacs.d/EmmyLua-LS-all.jar" ; default path
-       lua-lsp-emmy-java-path "java"                         ; default path
-       lua-lsp-emmy-enable-file-watchers t)                  ; enabled default
+       lsp-clients-emmy-lua-jar-path "~/.emacs.d/.cache/lsp/EmmyLua-LS-all.jar" ; default path
+       lsp-clients-emmy-lua-java-path "java" ; default path
+       lsp-enable-file-watchers t)           ; enabled default
+#+END_SRC
+
+To use the =lsp-lua-language-server= backend, please following [[https://github.com/sumneko/lua-language-server][lua-language-server]] to compile the standalone binnary,
```
[the remark this answers is not in the document]
Comment (sunlin7) [reply layers/+lang/lua/README.org]: resolved in new submition.

**Comment by sunlin7** [comment]
@lebensterben Thank you for your comments, I make changes and upload as https://github.com/syl20bnr/spacemacs/commit/42b71cc2efe03363fd7f6c72aa7b49da42c775c6, please review again.

File: layers/+lang/lua/README.org [hunk 11]
```diff
@@ -36,24 +36,34 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
 - =lua-mode= using emacs package
 
 *** LSP
-LSP support is provided via the [[file:../../+tools/lsp/README.org][LSP layer]], currently only [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua]] is supported.
+LSP support is provided via the [[file:../../+tools/lsp/README.org][LSP layer]], currently [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua]], [[https://github.com/sumneko/lua-language-server][lua-language-server]] and [[https://github.com/Alloyed/lua-lsp][lua-lsp]] are supported.
 
-To use the =lsp-emmy= backend, please download the [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua-LS-all.jar]] and put it into =~/.emacs.d/=,
-then set the layer variable lua-backend to =lsp-emmy=.
+To use the =lsp-emmy-lua= backend, please download the [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua-LS-all.jar]] and put it into =~/.emacs.d/=,
+then set the layer variable lua-backend to =lsp-emmy-lua=.
```
Comment (lebensterben) [inline layers/+lang/lua/README.org:48]: then -> and then

This is a grammar mistake. See https://www.wikiwand.com/en/Sentence_clause_structure#/Run-on_sentences

File: layers/+lang/lua/README.org [hunk 12]
```diff
@@ -36,24 +36,34 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
        lua-backend 'lsp-emmy
-       lua-lsp-emmy-jar-path "~/.emacs.d/EmmyLua-LS-all.jar" ; default path
-       lua-lsp-emmy-java-path "java"                         ; default path
-       lua-lsp-emmy-enable-file-watchers t)                  ; enabled default
+       lsp-clients-emmy-lua-jar-path "~/.emacs.d/.cache/lsp/EmmyLua-LS-all.jar" ; default path
+       lsp-clients-emmy-lua-java-path "java") ; default path
+#+END_SRC
+
+To use the =lsp-lua-language-server= backend, please follow [[https://github.com/sumneko/lua-language-server][lua-language-server]] to compile the standalone binnary,
+then set the layer variable lua-backend to =lsp-lua-language-server= and the variables for binnary and main entry, an example as below:and then set the layer variable =lua-backend= to =lsp-lua-language-server=, and the variables for binary and main entry such as the example bellow:
```
Comment (lebensterben) [inline layers/+lang/lua/README.org:61]: See my original comments.

There is are two grammar mistakes, and you haven't quoted the variable `lua-backend`.

File: layers/+lang/lua/funcs.el [hunk 13]
```diff
@@ -24,7 +24,11 @@
     "sr" 'lua-send-region
     "'" 'lua-show-process-buffer)
   (pcase lua-backend
-    (`lsp-emmy (spacemacs//lua-setup-lsp-emmy))))
+    ('lsp-emmy-lua (spacemacs//lua-setup-lsp-emmy-lua))
+    ('lsp-lua-language-server (spacemacs//lua-setup-lsp-lua-language-server))
+    ('lsp-lua-lsp (spacemacs//lua-setup-lsp-lua-lsp))
+    (_ (when (configuration-layer/layer-used-p 'lsp)
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:30]: The catuch-all case `_` should better throw an error instead.
For example,
```elisp
(_ (user-error "Unexpected value of `lua-backend', %s" lua-backend))
```

**Review by lebensterben — CHANGES_REQUESTED** [review]

File: layers/+lang/lua/README.org (hunk 11 above)
Comment (sunlin7) [reply layers/+lang/lua/README.org]: resolved in new submition.

File: layers/+lang/lua/funcs.el (hunk 13 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: Issue is resolved in new submition.

**Comment by sunlin7** [comment]
@lebensterben Thank you for comments and additional information. The new submition https://github.com/syl20bnr/spacemacs/commit/528888967b2df3905d5ae1f78f0163b52d2acc16 should be fixed all your mentioned points. 
Please help to review again.

File: layers/+lang/lua/funcs.el [hunk 14]
```diff
@@ -24,7 +24,12 @@
     "sr" 'lua-send-region
     "'" 'lua-show-process-buffer)
   (pcase lua-backend
-    (`lsp-emmy (spacemacs//lua-setup-lsp-emmy))))
+    ('lsp-emmy-lua (spacemacs//lua-setup-lsp-emmy-lua))
+    ('lsp-lua-language-server (spacemacs//lua-setup-lsp-lua-language-server))
+    ('lsp-lua-lsp (spacemacs//lua-setup-lsp-lua-lsp))
+    (_ (if (and (not lua-backend) (configuration-layer/layer-used-p 'lsp))
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:30]: This won't work. If `lua-backend` is simply nil, then it uses lua-mode and this is not a hard error.
The hard error is when the user supplied an unexpected value.

See my previous comments for my suggested changes.

**Comment by lebensterben** [comment]
> @lebensterben Thank you for comments and additional information. The new submition [5288889](https://github.com/syl20bnr/spacemacs/commit/528888967b2df3905d5ae1f78f0163b52d2acc16) should be fixed all your mentioned points.
> Please help to review again.

I updated my comments.
But still I don't like the calls to `make-obsolete-variable`.

File: layers/+lang/lua/funcs.el (hunk 14 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el]: In new submition https://github.com/syl20bnr/spacemacs/commit/3a56b390a7cd4c23d32c28186af00614bad225b2, it will be error for unexpect value, or try `lsp` when the `lsp-mode` avaiable.

**Comment by sunlin7** [comment]
Hi @lebensterben , a new submition https://github.com/syl20bnr/spacemacs/commit/3a56b390a7cd4c23d32c28186af00614bad225b2 to change the routine for default value and error value.
The variables `lua-lsp-emmy-java-path`, `lua-lsp-emmy-jar-path` and `lua-lsp-emmy-enable-file-watchers` should be remove in future. For `develop` branch provide weak back-compatible is acceptable for it always move forward.

**Comment by lebensterben** [comment]
I agree they are obsolete and should be removed.
But the thing is, we don't currently make variable or function obsolete in any layer.
And we don't have a clear policy whether `develop` branch should include breaking changes.

**Comment by sunlin7** [comment]
@lebensterben According your comments, I remove the obsolate variables for lua layer, submission is https://github.com/syl20bnr/spacemacs/commit/71be783e3aac42f7836de334ca28ea29eb2a516e. 
Please review again. Thanks.

**Comment by sunlin7** [comment]
@lebensterben Could you please review the last submission ? Thanks.

**Comment by sunlin7** [comment]
Can anyone help to review this change please?

**Comment by sunlin7** [comment]
@smile13241324

File: layers/+lang/lua/README.org [hunk 15]
```diff
@@ -36,24 +36,34 @@ In order to enable code linting, install [[https://github.com/mpeterv/luacheck][
...
 - =lua-mode= using emacs package
 
 *** LSP
-LSP support is provided via the [[file:../../+tools/lsp/README.org][LSP layer]], currently only [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua]] is supported.
+LSP support is provided via the [[file:../../+tools/lsp/README.org][LSP layer]], currently [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua]], [[https://github.com/sumneko/lua-language-server][lua-language-server]] and [[https://github.com/Alloyed/lua-lsp][lua-lsp]] are supported.
 
-To use the =lsp-emmy= backend, please download the [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua-LS-all.jar]] and put it into =~/.emacs.d/=,
-then set the layer variable lua-backend to =lsp-emmy=.
+To use the =lsp-emmy-lua= backend, please download the [[https://github.com/EmmyLua/EmmyLua-LanguageServer][EmmyLua-LS-all.jar]] and put it into =~/.emacs.d/=,
+and then set the layer variable lua-backend to =lsp-emmy-lua=.
```
Comment (lebensterben) [inline layers/+lang/lua/README.org:48]: lua-backend -> =lua-backend=

File: layers/+lang/lua/funcs.el [hunk 16]
```diff
@@ -24,7 +24,13 @@
     "sr" 'lua-send-region
     "'" 'lua-show-process-buffer)
   (pcase lua-backend
-    (`lsp-emmy (spacemacs//lua-setup-lsp-emmy))))
+    ('lsp-emmy-lua (spacemacs//lua-setup-lsp-emmy-lua))
+    ('lsp-lua-language-server (spacemacs//lua-setup-lsp-lua-language-server))
+    ('lsp-lua-lsp (spacemacs//lua-setup-lsp-lua-lsp))
+    (_ (if lua-backend
+           (user-error "Unexpected value of `lua-backend', %s" lua-backend)
+         (when (configuration-layer/layer-used-p 'lsp)
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:32]: The last `when` clause results in undefined behaviour that which of the lua LSP client will be used is not clear.

```elisp
(setq-local lsp-disabled-clients
            '(emmy-lua lsp-lua-language-server lsp-lua-lsp))
(if (member lua-backend lsp-disabled-clients)
    (delq lua-backend lsp-disabled-clients)
  (user-error "Unexpected value of `lua-backend', %s" lua-backend))
```

File: layers/+lang/lua/funcs.el [hunk 17]
```diff
@@ -33,13 +39,19 @@
...

 ;; LSP Lua
-(defun spacemacs//lua-setup-lsp-emmy ()
-  "Setup LSP Lua."
-  (when lua-lsp-emmy-java-path
-    (setq lsp-clients-emmy-lua-java-path lua-lsp-emmy-java-path))
-  (when lua-lsp-emmy-jar-path
-    (setq lsp-clients-emmy-lua-jar-path (expand-file-name lua-lsp-emmy-jar-path)))
-  (setq lsp-enable-file-watchers lua-lsp-emmy-enable-file-watchers)
+(defun spacemacs//lua-setup-lsp-emmy-lua ()
```
Comment (lebensterben) [inline layers/+lang/lua/funcs.el:42]: These three functions can be combined.
For setting `lsp-disable-clients`, these actually are not needed.

See comments above.

**Review by lebensterben — CHANGES_REQUESTED** [review]
Please see my comments.

File: layers/+lang/lua/funcs.el (hunk 16 above)
Comment (sunlin7) [reply layers/+lang/lua/funcs.el:32]: @lebensterben The last `when` clause will fallback to default behavior for user who just enable `lsp` and `lua` layer without any more configurations, `lsp` will try detect the `emmy-lua`, `lsp-lua-language-server`, `lsp-lua-lsp` in PATH or not.

Comment (lebensterben) [reply layers/+lang/lua/funcs.el:32]: That's exactly what undefined behavior is.

Comment (sunlin7) [reply layers/+lang/lua/funcs.el:32]: Understood your points, and I also understood `lsp-mode` design that less configuration, more automation.

Comment (sunlin7) [reply layers/+lang/lua/funcs.el:32]: I'll add more documents in `Readme.org` to make it exactly how `lsp-mode` work in `lua` layer.

Comment (lebensterben) [reply layers/+lang/lua/funcs.el:32]: Disagree.
If something is not working as expected, users should be clearly notified.

It's possible the user has multiple clients installed and he has a typo in dotspacemacs. Then your current commit would rely on lsp-mode to figure out which client to use, and that may not be the one user wanted.

Since there's no message about the invalid config, the use will be confused why lsp-mode is not using the clien@t he specified.

Comment (sunlin7) [reply layers/+lang/lua/funcs.el:32]: When user give wrong configuration, current code will give `user-error` message. 
When user just enable `lsp` and `lua` layer without other configuration, what should Spacemacs do? Spacsmacs should have default behavior. 
The default behavior can do warn with a message, or, try its best to find a working set.

Comment (sunlin7) [reply layers/+lang/lua/funcs.el:32]: Many Emacs user begin their Emacs experience from Spacemacs than other emacs framework (eg: doom-emacs) for Spacemacs is simple and elegant, especially it's usable with the default configuration.
Exactly configurations are necessary for experts; while default configurations keep friendly for freshman is worth also.

Comment (lebensterben) [reply layers/+lang/lua/funcs.el:32]: But here's a possible scenario why the "default behaviour" actually results in unexpected behaviours.

1. Suppose multiple lua clients, `a` and `b`, are installed under two different locations, `A` and `B`.
2. Suppose, for example, uses `bash`.
   - `.bash_profile` is loaded for login shell only, and it has `export PATH=A`.
   - `.bashrc` is loaded for interactive shell only, and it has `export PATH=$PATH:B`.
3. Suppose user uses a graphical login manager, and environment variables sets in
   `.bash_profile` is not inherited by any child process of the login manager.
   - When user opens a terminal, it runs a non-login, interactive `bash` shell. And `PATH=B`, hence `b` is found first and used as the LSP client.
   - When user connects to this host via `ssh`, it runs a login, interactive shell. So `PATH=A:B`, and then `a` is found first and used as the LSP client.

In this simple scenario, `lsp-mode` would use different LSP client, without showing any explicit message.

Since each LSP client may provide different [capabilities](https://docs.microsoft.com/en-us/visualstudio/extensibility/language-server-protocol?view=vs-2019#capabilities), some commands in Emacs may work differently. Then it could be a huge pain for the user to figure out why sometimes a command doesn't work as expected.

**Comment by smile13241324** [comment]
Looks like discussion is still going on here, I will later have a look at this PR, in the meantime you should have a look at how multiple LSP backends are handled in `python-layer`. Especially the default backend choice must be handled in the same way to allow consistent behaviour among all LSP supporting layers.

**Comment by smile13241324** [comment]
I have manually merged this PR and refactored the LSP integration in this layer to match the standard procedures in the rest of Spacemacs.

**Comment by sunlin7** [comment]
Be Commited as ca0fe33c8d417ab1cfba86a1f76bed1d1ad72c70 and revised in fc2aa2601c8b23c4f53ebd27dd96eaca06dab9b2.
Thanks @smile13241324 .
